### PR TITLE
fix nullptr dereference in prune_requests_older_than

### LINKED ISSUE #2007: SEGFAULT in prune_requests_older_than() if pruned_requests are not provided
Segfault happens here: 
https://github.com/ros2/rclcpp/blob/dab9f5e0a331b9694c4b5aceb8263d49aa052ecb/rclcpp/include/rclcpp/client.hpp#L772

The documentation says 
```
[inout] pruned_requests Removed requests id will be pushed to the vector if a pointer is provided.
```
So I assume no pointer is a valid input.

Please, fix it, or tell me which branch to submit the fix, so all supported versions were updated.

### PR DESCRIPTION
**Pull request by akela1101**
Fixes #2007

### CODE REVIEW

**Review by alsora — APPROVED** [review]

**Review by fujitatomoya — APPROVED** [review]

**Review by clalancette — CHANGES_REQUESTED** [review]
This function is buggy as you point out, so thanks for the fix.

That said, the reason we didn't know it was wrong was that we don't have a test for this function.  Could you add two tests, one of which has `pruned_requests` as nullptr (or just not passed), and one of which has `pruned_requests` as a pointer?

(also, completely orthogonally, why is `pruned_requests` a pointer and not a reference?  Maybe @ivanpauno can chime in since he added this API)

**Comment by ivanpauno** [comment]
> (also, completely orthogonally, why is pruned_requests a pointer and not a reference? Maybe @ivanpauno can chime in since he added this API)

It's an optional output argument, so the vector isn't used if `nullptr` is provided.
References cannot be nullptr, so they cannot be used in this case.
For some reason I forgot the obvious `if()` in the original PR.

**Comment by akela1101** [comment]
@clalancette Sorry for waiting. I tried to add 2 tests, please check.

The way I tested them locally was:
```
colcon build --packages-select rclcpp
colcon test --packages-select rclcpp --ctest-args -R test_client
colcon test-result --verbose
```

**Review by alsora — APPROVED** [review]

File: rclcpp/test/rclcpp/test_client.cpp [hunk 1]
```diff
@@ -327,6 +327,26 @@ TEST_F(TestClientWithServer, test_client_remove_pending_request) {
   EXPECT_TRUE(client->remove_pending_request(future));
 }
 
+TEST_F(TestClientWithServer, prune_requests_older_than_no_pruned) {
+  auto client = node->create_client<test_msgs::srv::Empty>(service_name);
+  auto request = std::make_shared<test_msgs::srv::Empty::Request>();
+  auto future = client->async_send_request(request);
+  auto time = std::chrono::system_clock::now() + 1s;
+
+  EXPECT_EQ(size_t(1), client->prune_requests_older_than(time));
```
Comment (ivanpauno) [inline rclcpp/test/rclcpp/test_client.cpp:336]: I would rather use:

```suggestion
  EXPECT_EQ(1u, client->prune_requests_older_than(time));
```

or maybe `size_t{1}`.

File: rclcpp/test/rclcpp/test_client.cpp [hunk 2]
```diff
@@ -327,6 +327,26 @@ TEST_F(TestClientWithServer, test_client_remove_pending_request) {
...
+
+TEST_F(TestClientWithServer, prune_requests_older_than_with_pruned) {
+  auto client = node->create_client<test_msgs::srv::Empty>(service_name);
+  auto request = std::make_shared<test_msgs::srv::Empty::Request>();
+  auto future = client->async_send_request(request);
+  auto time = std::chrono::system_clock::now() + 1s;
+
+  std::vector<int64_t> pruned_requests;
+  EXPECT_EQ(size_t(1), client->prune_requests_older_than(time, &pruned_requests));
+  EXPECT_EQ(size_t(1), pruned_requests.size());
```
Comment (ivanpauno) [inline rclcpp/test/rclcpp/test_client.cpp:347]:
```suggestion
  EXPECT_EQ(1u, client->prune_requests_older_than(time, &pruned_requests));
  EXPECT_EQ(1u, pruned_requests.size());
```

**Review by ivanpauno — APPROVED** [review]
lgtm!

**Comment by akela1101** [comment]
> 1u

nice suggestion, fixed

**Review by ivanpauno — APPROVED** [review]

**Review by fujitatomoya — APPROVED** [review]

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17727/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12269/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18367/)

File: rclcpp/test/rclcpp/test_client.cpp [hunk 3]
```diff
@@ -327,6 +327,26 @@ TEST_F(TestClientWithServer, test_client_remove_pending_request) {
...
+
+TEST_F(TestClientWithServer, prune_requests_older_than_with_pruned) {
+  auto client = node->create_client<test_msgs::srv::Empty>(service_name);
+  auto request = std::make_shared<test_msgs::srv::Empty::Request>();
+  auto future = client->async_send_request(request);
+  auto time = std::chrono::system_clock::now() + 1s;
+
+  std::vector<int64_t> pruned_requests;
+  EXPECT_EQ(1u, client->prune_requests_older_than(time, &pruned_requests));
+  EXPECT_EQ(1u, pruned_requests.size());
```
Comment (iuhilnehc-ynos) [inline rclcpp/test/rclcpp/test_client.cpp:347]:
```suggestion
  ASSERT_EQ(1u, pruned_requests.size());
  EXPECT_EQ(future.request_id, pruned_requests[0]);
```

**Comment by ivanpauno** [comment]
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18373/)

**Comment by ivanpauno** [comment]
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18405/)

**Comment by ivanpauno** [comment]
@mergifyio backport humble

**Pull request merged**